cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+type-dependent lookups in the compiler, which can complicate type
+inference). For example, given the above `MyType` definitions, we
+automatically have the following constructors:
+
+```lisp
+
+    ;; These definitions are implicit and do not need to be written (doing
+    ;; so is a compile-time error, actually). We write them here just to
+    ;; show what they would look like.
+    
+    (decl MyType.A (u32 u32) MyType)
+    (decl MyType.B (u32) MyType)
+    (decl MyType.C () MyType)
+    
+    (decl MyType2.A () MyType2)
+```
+
+### Why Types?
+
+For terms that are not enum variants, the notion that a term "has a
+type" is somewhat foreign to a classical term-rewriting system. In
+most formal symbolic systems, the terms are manipulated as opaque
+sequences or trees of symbols; they have no inherent meaning other
+than what the user implicitly defines with the given rules. What does
+it mean for a term to "have a type" when it is just data? Or, said
+another way: why isn't the type of `Term2` just `Term2`?
+
+The types come into play when we define *rules*: one term of type `T`
+can only be rewritten into another term of type `T`, and when a
+parameter has a certain type, only subterms with that type can
+appear. Without explicit types on terms and their parameters, any term
+could be rewritten to any other, or substituted in as a parameter, and
+there is thus a kind of dynamic typing about which the programmer must
+have some caution. In most applications of a term-rewriting system,
+there is already some de-facto "schema": some parameter of a term
+representing a machine instruction can only take on one of a few
+subterms (representing, say, different addressing modes). ISLE's types
+just make this explicit.
+
+Thus, the first answer to "why types" is that they enforce a schema on
+the terms, allowing the programmer to have stronger well-formed-data
+invariants.
+
+The second reason is that the types are an integral part of the
+compilation-to-Rust strategy: every constructor actually does evaluate
+to a Rust value of the given "return value" type, given actual Rust
+values for its parameters of the appropriate parameter types. We will
+see more on this below.
+
+### Well-Typed Rules and Type Inference
+
+Now that we understand how to define types, let's examine in more
+detail how they are used to verify that the pattern and rewrite
+expression of a rule have the same type.
+
+ISLE uses a simple unidirectional type-inference algorithm that
+propagates type information through the pattern, resulting in a "type
+environment" that specifies the type for each captured variable, and
+then uses this to typecheck the rewrite expression. The result of this
+is that types are almost completely inferred, and are only annotated
+in a few places (`let` bindings specifically).
+
+The typing rules for patterns in ISLE are:
+
+* At the root of the pattern, we require that a *constructor* pattern
```

Comment:
Can wildcards be used as constructions if they are nested? I.e. `(A (_ 1 x))`?

Reply:
No, a constructor form has to have a concrete constructor name; no polymorphism in that dimension. Added clarification, thanks.